core/src/main/scala/org/apache/spark/deploy/security/KafkaTokenUtil.scala

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.security
+
+import java.text.SimpleDateFormat
+import java.util.Properties
+
+import org.apache.hadoop.io.Text
+import org.apache.hadoop.security.token.{Token, TokenIdentifier}
+import org.apache.hadoop.security.token.delegation.AbstractDelegationTokenIdentifier
+import org.apache.kafka.clients.CommonClientConfigs
+import org.apache.kafka.clients.admin.{AdminClient, CreateDelegationTokenOptions}
+import org.apache.kafka.common.config.SaslConfigs
+import org.apache.kafka.common.security.token.delegation.DelegationToken
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+
+private[spark] object KafkaTokenUtil extends Logging {
+  private[spark] val TOKEN_KIND = new Text("KAFKA_DELEGATION_TOKEN")
+  private[spark] val TOKEN_SERVICE = new Text("kafka.server.delegation.token")
+
+  private[spark] class KafkaDelegationTokenIdentifier extends AbstractDelegationTokenIdentifier {
+    override def getKind: Text = TOKEN_KIND;
+  }
+
+  private[security] def obtainToken(sparkConf: SparkConf): (Token[_ <: TokenIdentifier], Long) = {
+    val adminClient = AdminClient.create(createAdminClientProperties(sparkConf))
+    val createDelegationTokenOptions = new CreateDelegationTokenOptions()
+    val createResult = adminClient.createDelegationToken(createDelegationTokenOptions)
+    val token = createResult.delegationToken().get()
+    printToken(token)
+
+    (new Token[KafkaDelegationTokenIdentifier](
+      token.tokenInfo.tokenId.getBytes,
+      token.hmacAsBase64String.getBytes,
+      TOKEN_KIND,
+      TOKEN_SERVICE
+    ), token.tokenInfo.expiryTimestamp)
+  }
+
+  private[security] def createAdminClientProperties(sparkConf: SparkConf): Properties = {
+    val adminClientProperties = new Properties
+
+    val bootstrapServers = sparkConf.get(KAFKA_BOOTSTRAP_SERVERS)
+    require(bootstrapServers.nonEmpty, s"Tried to obtain kafka delegation token but bootstrap " +
+      "servers not configured.")
+    adminClientProperties.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers.get)
+
+    val protocol = sparkConf.get(KAFKA_SECURITY_PROTOCOL)
+    adminClientProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, protocol)
+    if (protocol.endsWith("SSL")) {
+      logInfo("SSL protocol detected.")
+      sparkConf.get(KAFKA_TRUSTSTORE_LOCATION).foreach { truststoreLocation =>
+        adminClientProperties.put("ssl.truststore.location", truststoreLocation)
+      }
+      sparkConf.get(KAFKA_TRUSTSTORE_PASSWORD).foreach { truststorePassword =>
+        adminClientProperties.put("ssl.truststore.password", truststorePassword)
+      }
+    } else {
+      logWarning("Obtaining kafka delegation token through plain communication channel. Please " +
+        "consider the security impact.")
+    }
+
+    // There are multiple possibilities to log in:
```

Comment:
If I understand this correctly, this code is trying to log in again, right? Two questions:

- why is that necessary?
- won't this fail if the user only has a TGT (i.e. login via kinit)?

I was expecting that you could reuse the already available login information to talk to the Kafka broker here and get delegation tokens. For example, you can do this to change the JVM's security context to the current user, if it's not already set:

```
org.apache.hadoop.security.UserGroupInformation.getCurrentUser().doAs(
  new java.security.PrivilegedExceptionAction[Unit] { 
    override def run(): Unit = {
      println(javax.security.auth.Subject.getSubject(
        java.security.AccessController.getContext()))
    }
  }
)

// Exiting paste mode, now interpreting.

Subject:
        Principal: UnixPrincipal: blah
        Principal: UnixNumericUserPrincipal: 2000
        Principal: UnixNumericGroupPrincipal [Primary Group]: 2000
 ```

Wouldn't that be enough here? That `Subject` has a TGT, so that should be enough to talk to Kafka. Having to do yet another login just for Kafka feels a bit weird.

Reply:
you are right: the user shouldn't need to be logging in again, as normal end-user use is to be kinited in, rather than giving a keytab

Reply:
This code doesn't try to log in but provides dynamic jaas configuration to kafka. In the original implementation only jaas was possible now I've extended it to make things easier for the user.

> why is that necessary?

This eliminates the need to provide file based jaas configuration.
Old way: https://github.com/gaborgsomogyi/spark-structured-secure-kafka-app
New way: https://github.com/gaborgsomogyi/spark-structured-secure-dt-kafka-app

> won't this fail if the user only has a TGT (i.e. login via kinit)?

It works. TGT can be provided to kafka by providing a file based jaas file with content like:
```
useTicketCache=true
ticketCache=<path>
```

Please see how kafka client authentication works: https://cwiki.apache.org/confluence/display/KAFKA/KIP-85%3A+Dynamic+JAAS+configuration+for+Kafka+clients

Reply:
JAAS configuration files control how to *log in*. That's what the JAAS API does; it's an abstraction to log into different auth services (much like PAM). So if you're using JAAS here, you're asking the Kafka API to log in. (And btw, even though it's not explicit, that's basically what the Kafka doc you linked to is saying.)

If you give it a tgt-based config file, you're telling JAAS to load the credentials from the ticket cache.

If you give it a keytab-based config file, you're telling JAAS to use the keytab to authenticate against the KDC.

etc etc

And that's not needed at this point, because that has already been done. Once JAAS logs in, what happens is that you have a `Subject` in the current security context with the appropriate credentials. As my example above shows, that's already done by the Hadoop APIs that Spark uses.

So my question is why can't the Kafka libraries use the already logged in `Subject`.

Reply:
I've just double checked my knowledge and this case kafka looks for JAAS entry but will not find it so it throws this exception:
```
Caused by: java.lang.IllegalArgumentException: Could not find a 'KafkaClient' entry in the JAAS configuration. System property 'java.security.auth.login.config' is not set
```

Reply:
Where does that exception come from? The message by itself doesn't help.

And that still does not answer my question. The fact it's even talking about JAAS configs tells me that it's still trying to log in. My question is whether there's an API in Kafka that means "just connect to the server with my existing credentials".

At this point in the code, you are already logged in. You already have credentials. Kafka should not be forcing you to go through all this again.

It may be that the answer is "Kafka doesn't have this API" in which case a bug in Kafka should be filed, and we may have to live with this until that is fixed. But before going forward here, I'd really like to be sure that is really the case.

Reply:
OK, speaking with them...